File: java/dsa/algorithms/trie/add-search.java
class WordDictionary {
    private class Node {
        Map<Character, Node> children = new HashMap<>();
        boolean isEnd = false;
    }

    private Node root;

    public WordDictionary() { // @viz:init
        root = new Node(); // @viz:init
    }

    public void addWord(String word) { // @viz:add_start
        Node curr = root; // @viz:add_start
        for (char c : word.toCharArray()) { // @viz:add_loop
            if (!curr.children.containsKey(c)) { // @viz:add_check
                curr.children.put(c, new Node()); // @viz:add_new
            }
            curr = curr.children.get(c); // @viz:add_next
        }
        curr.isEnd = true; // @viz:add_end
    }

    public boolean search(String word) { // @viz:search_start
        return dfs(root, 0, word); // @viz:search_start
    }

    private boolean dfs(Node node, int index, String word) { // @viz:dfs_start
        if (index == word.length()) { // @viz:dfs_base_check
            return node.isEnd; // @viz:dfs_base
        }
        
        char c = word.charAt(index); // @viz:dfs_char
        if (c != '.') { // @viz:dfs_not_dot_check @viz:dfs_not_dot
            if (!node.children.containsKey(c)) { // @viz:dfs_check_fail
                return false; // @viz:dfs_check_fail
            }
            return dfs(node.children.get(c), index + 1, word); // @viz:dfs_recurse
        }
        
        for (Node child : node.children.values()) { // @viz:dfs_dot_loop
            if (dfs(child, index + 1, word)) { // @viz:dfs_dot_recurse
                return true; // @viz:dfs_dot_found
            }
        }
        return false; // @viz:dfs_fail
    }
}

File: python/dsa/algorithms/trie/add-search.py
class WordDictionary:
    def __init__(self): # @viz:init
        self.root = { # @viz:init
            "children": {}, # @viz:init
            "isEnd": False # @viz:init
        } # @viz:init

    def addWord(self, word: str) -> None: # @viz:add_start
        curr = self.root # @viz:add_start
        for char in word: # @viz:add_loop
            if char not in curr["children"]: # @viz:add_check
                curr["children"][char] = { # @viz:add_new
                    "children": {}, # @viz:add_new
                    "isEnd": False # @viz:add_new
                } # @viz:add_new
            curr = curr["children"][char] # @viz:add_next
        curr["isEnd"] = True # @viz:add_end

    def search(self, word: str) -> bool: # @viz:search_start
        return self.dfs(self.root, 0, word) # @viz:search_start

    def dfs(self, node, index, word): # @viz:dfs_start
        if index == len(word): # @viz:dfs_base_check
            return node["isEnd"] # @viz:dfs_base
        
        char = word[index] # @viz:dfs_char
        if char != '.': # @viz:dfs_not_dot_check @viz:dfs_not_dot
            if char not in node["children"]: # @viz:dfs_check_fail
                return False # @viz:dfs_check_fail
            return self.dfs( # @viz:dfs_recurse
                node["children"][char], # @viz:dfs_recurse
                index + 1, # @viz:dfs_recurse
                word # @viz:dfs_recurse
            ) # @viz:dfs_recurse
            
        for key in node["children"]: # @viz:dfs_dot_loop
            if self.dfs(node["children"][key], index + 1, word): # @viz:dfs_dot_recurse
                return True # @viz:dfs_dot_found
        return False # @viz:dfs_fail